src/elli_http.erl

```diff
@@ -167,38 +167,46 @@ handle_response(Req, Buffer, {file, ResponseCode, UserHeaders,
 
     ResponseHeaders = [connection(Req, UserHeaders) | UserHeaders],
 
-    Size  = elli_util:file_size(Filename),
-
-    t(send_start),
-    case elli_util:normalize_range(Range, Size) of
-        undefined ->
-            send_file(Req, ResponseCode, [{<<"Content-Length">>, Size} |
-                                          ResponseHeaders],
-                      Filename, {0, 0});
-        {Offset, Length} ->
-            ERange = elli_util:encode_range({Offset, Length}, Size),
-            send_file(Req, 206, lists:append(ResponseHeaders,
-                                             [{<<"Content-Length">>, Length},
-                                              {<<"Content-Range">>, ERange}]),
-                      Filename, {Offset, Length});
-        invalid_range ->
-            ERange = elli_util:encode_range(invalid_range, Size),
-            send_response(Req, 416,
-                          lists:append(ResponseHeaders,
-                                       [{<<"Content-Length">>, 0},
-                                        {<<"Content-Range">>, ERange}]),
-                          [])
-    end,
-    t(send_end),
-
-    t(request_end),
-    handle_event(Mod, request_complete,
-                 [Req, ResponseCode, ResponseHeaders, <<>>, {get_timings(),
-                                                             get_sizes()}],
-                 Args),
+    case elli_util:file_size(Filename) of
+        {error, FileError} ->
+            handle_event(Mod, file_error, [FileError], Args),
+            send_server_error(Req#req.socket),
+            elli_tcp:close(Req#req.socket),
+            exit(normal);
+        Size ->
+            t(send_start),
+            case elli_util:normalize_range(Range, Size) of
```

Comment:
In a future release, we should try to pull these nested `case` expressions out to (a) helper function(s), but I think this is fine for now.